adr/0005-convention-based-logging-classes.md

```diff
@@ -0,0 +1,111 @@
+
+> *XXX = Severity
+> X*XX = Hamlet Framework Component
+> XX** = Unique log code
+
+For the first digit, existing log message severities would be used. Some digits are currently listed as "spare" severity levels which should simmilarly be spares under this system.
+
+For the second digit, the Hamlet Framework Components would be assigned as follows:
+
+> Generic/Unaccounted For Error - 0
+> Freemarker Wrapper (Java) - 1
+> Engine - 2
+> Engine Plugins - 3
+> Executor - bash - 4
+> Executor - python - 5
+
+### Use an Inherritance-based Class Approach
+
+Design a layered log message system where a log message inherrits its severity and origin from parent classes, which themseles contribute greater detail to the message.
+
+> Severity
+>   Origin
+>     Message
+
+### Centralised Code Definitions
+
+Assuming the selection of the numerical convention-based approach, a central definition of the codes is developed. Each Hamlet Framework Component would then be updated with a mechanism to interpret and use them.
+
+### Per-Component Code Range
+
+Assuming the selection of the numerical convention-based approach, digits 3 & 4 would then provide each Framework Component a range of 99 possible unique log messages.
+
+Each Component would create a "dictionary" of this range and their associated log messages. When a new message is written, either an existing digit is used again (assuming matching cause of the log message) or the next unallocated number in the 00-99 range is added to the dictionary.
+
+Each Component would then simply append the 2-digit-code to the message's severity & it's own assigned Framework Component digit to form the logging code.
+
+## Decision Outcome
+
+Chosen option:  _"Use Logging Codes Followig a Numerical Convention"_ with _"Per-Component Code Range"_
+
+It was deemed easier to implement and there seemed little benefit and greater complexity to implementing an inherritance-based approach. 
+
+A code range allows each component to implement the definitions in the way that most makes sense to it.
+
+## Pros and Cons of the Options
+
+### Use Logging Codes Followig a Numerical Convention
+
+* Good, because its easy to maintain
+* Good, because each Hamlet Framework Component has easily understood log message boundaries
+* Good, because a good deal of information can be determined from an log message code on its own
+* Good, because it standardises the approach to log messages across Hamlet Deploy
+* Bad, because there are no rules around how and when an existing code should be reused and when a new one assigned
+
+### Use an Inherritance-based Class Approach
+
+* Good, because inherritance will enable the greatest re-use of log codes
+* Good, because a log message can contain varying scopes of detail - ie. a known limitiation of the Engine for instance could be inherrited by a child, and a subsequent error will not only indicate the cause but provide background on that limitation
```

Comment:
```suggestion
* Good, because a log message can contain varying scopes of detail - ie. a known limitation of the Engine for instance could be inherited by a child, and a subsequent error will not only indicate the cause but provide background on that limitation
```